src/js/helpers/nodesHelper.js

```diff
@@ -14,65 +14,81 @@
  * limitations under the License.
  */
 
-export const knotxNodes = () => {
-  const rootElem = document.body;
-  const COMMENTS = [];
-  const COMMENT_NODE_CODE = 8;
-  const HTML_NODE_CODE = 1;
-  // Fourth argument, which is actually obsolete according to the DOM4 standard, is required in IE 11
-  const iterator = document.createNodeIterator(
-    rootElem,
-    NodeFilter.SHOW_ALL,
-  );
-
-  let curNode;
-  let isBetweenComments = false;
-  let curComment = '';
-  // eslint-disable-next-line no-cond-assign
-  while ((curNode = iterator.nextNode())) {
-    if (
-      curNode.nodeType === COMMENT_NODE_CODE
-          && curNode.data.indexOf('data-knotx-id') !== -1
-    ) {
-      isBetweenComments = !isBetweenComments;
-      curComment = isBetweenComments ? curNode.data.match(/"([^']+)"/)[1] : '';
+const isFragmentBoundary = (node) => node.nodeType === Node.COMMENT_NODE && node.data.includes('data-knotx-id');
+
+const isCorrectFragmentNode = (node) => node.nodeType === Node.ELEMENT_NODE
+  || (node.nodeType === Node.TEXT_NODE && node.textContent.trim());
+
+const getKnotxId = (node) => node.data.match(/data-knotx-id="(.*)"/)[1];
+
+const isDebugDataNode = (node, fragmentId) => node && node.tagName === 'SCRIPT' && node.dataset.knotxId === fragmentId;
+
+const findBoundaries = (root) => {
+  const iterator = document.createNodeIterator(root, NodeFilter.SHOW_ALL);
+  const boundaries = [];
+  let node = iterator.nextNode();
+
+  while (node) {
+    if (isFragmentBoundary(node)) {
+      boundaries.push(node);
     }
 
-    if (
-      isBetweenComments
-          && curNode.nodeType !== COMMENT_NODE_CODE
-          && curNode.nodeType === HTML_NODE_CODE
-    ) {
-      if (!COMMENTS.length) {
-        curNode.dataset.knotxId = curComment;
-        COMMENTS.push(curNode);
-      } else if (!COMMENTS[COMMENTS.length - 1].contains(curNode)) {
-        curNode.dataset.knotxId = curComment;
-        COMMENTS.push(curNode);
+    node = iterator.nextNode();
+  }
+
+  return boundaries;
+};
+
+const findFragmentNodes = (root) => {
+  const boundaries = findBoundaries(root);
+  const fragments = new Map();
```

Comment:
If you want to have unique elements that will be Set

Reply:
I cannot easily find the element I'm looking for in a set without iterating over it, which is essential in this algorithm. That's why I map objects to their IDs in a `Map`. My first thought was to just use a plain object, but I think `Map` is designed for this very case